presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RuntimeReorderJoinSides.java

```diff
@@ -0,0 +1,163 @@
+import com.facebook.presto.spi.PrestoWarning;
+import com.facebook.presto.spi.plan.PlanNode;
+import com.facebook.presto.spi.plan.TableScanNode;
+import com.facebook.presto.spi.relation.VariableReferenceExpression;
+import com.facebook.presto.sql.planner.iterative.Rule;
+import com.facebook.presto.sql.planner.plan.ExchangeNode;
+import com.facebook.presto.sql.planner.plan.JoinNode;
+import com.google.common.collect.ImmutableSet;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+
+import static com.facebook.presto.SystemSessionProperties.getTaskConcurrency;
+import static com.facebook.presto.spi.StandardWarningCode.PERFORMANCE_WARNING;
+import static com.facebook.presto.sql.planner.plan.ExchangeNode.Scope.LOCAL;
+import static com.facebook.presto.sql.planner.plan.ExchangeNode.gatheringExchange;
+import static com.facebook.presto.sql.planner.plan.ExchangeNode.systemPartitionedExchange;
+import static com.facebook.presto.sql.planner.plan.Patterns.join;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.collect.Streams.stream;
+import static com.google.common.graph.Traverser.forTree;
+import static java.lang.String.format;
+
+public class RuntimeReorderJoinSides
+        implements Rule<JoinNode>
+{
+    private final Logger log = Logger.get(RuntimeReorderJoinSides.class);
+
+    private static final Pattern<JoinNode> PATTERN = join();
+
+    @Override
+    public Pattern<JoinNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Result apply(JoinNode joinNode, Captures captures, Context context)
+    {
+        try {
+            // Early exit if the leaves of the joinNode subtree do not include table scan.
+            if (stream(forTree(node -> context.getLookup().resolve((PlanNode) node).getSources()).depthFirstPreOrder(joinNode))
+                    .noneMatch(child -> context.getLookup().resolve((PlanNode) child) instanceof TableScanNode)) {
+                return Result.empty();
+            }
+            StatsProvider statsProvider = context.getStatsProvider();
+            double leftOutputSizeInBytes = statsProvider.getStats(joinNode.getLeft()).getOutputSizeInBytes(joinNode.getLeft().getOutputVariables());
+            double rightOutputSizeInBytes = statsProvider.getStats(joinNode.getRight()).getOutputSizeInBytes(joinNode.getRight().getOutputVariables());
+
+            // Only swap join sides when outputRowCount is available for both sides.
+            if (!Double.isNaN(leftOutputSizeInBytes) && !Double.isNaN(rightOutputSizeInBytes) && rightOutputSizeInBytes > leftOutputSizeInBytes) {
+                JoinNode swapped = joinNode.flipChildren();
+
+                PlanNode newLeft = swapped.getLeft();
+                PlanNode resolvedNewLeft = context.getLookup().resolve(newLeft);
+                // Remove unnecessary LocalExchange in the current probe side. If the immediate left child (new probe side) of the join node
```

Comment:
I don't love the implicit dependency here with AddLocalExchanges because if something changes there, nobody is going to know to update this. Maybe it would be better if we just removed the local exchanges and then ran AddLocalExchanges again after this.

Reply:
Discussed offline. 

AddLocalExchanges will transform all Unions into LocalExchanges, and this is why it is not viable just remove all local exchanges and add them back later on by running this AddLocalExchanges optimizer.

Instead, we decide to let it be here, and add a sanity checker to report an error if AddLocalExchanges logic later on evolves to be different from our adjustments of the localExchanges here.